drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+	}
+
+	if (whd_wifi_init_ap(airoc_ap_if, &ssid, security, (const uint8_t *)params->psk,
+			     params->psk_length, channel) != 0) {
+		LOG_ERR("Failed to init whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	if (whd_wifi_start_ap(airoc_ap_if) != 0) {
+		LOG_ERR("Failed to start whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	/* set event handler */
+	if (whd_management_set_event_handler(airoc_ap_if, ap_link_events, ap_link_events_handler,
+					     NULL, &ap_event_handler_index) != 0) {
+		whd_wifi_stop_ap(airoc_ap_if);
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	data->is_ap_up = 1;
+
+	k_sem_give(&data->sema_common);
+	return 0;
+}
+
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+static int airoc_mgmt_wifi_stats(const struct device *dev, struct net_stats_wifi *stats)
+{
+	struct airoc_wifi_data *data = dev->data;
+
+	stats->bytes.received = data->stats.bytes.received;
+	stats->bytes.sent = data->stats.bytes.sent;
+	stats->pkts.rx = data->stats.pkts.rx;
+	stats->pkts.tx = data->stats.pkts.tx;
+	stats->errors.rx = data->stats.errors.rx;
+	stats->errors.tx = data->stats.errors.tx;
+	stats->broadcast.rx = data->stats.broadcast.rx;
+	stats->broadcast.tx = data->stats.broadcast.tx;
+	stats->multicast.rx = data->stats.multicast.rx;
+	stats->multicast.tx = data->stats.multicast.tx;
+	stats->sta_mgmt.beacons_rx = data->stats.sta_mgmt.beacons_rx;
+	stats->sta_mgmt.beacons_miss = data->stats.sta_mgmt.beacons_miss;
+
+	return 0;
+}
+#endif
+
+static int airoc_mgmt_ap_disable(const struct device *dev)
+{
+	int ret = 0;
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	whd_wifi_deregister_event_handler(airoc_ap_if, ap_event_handler_index);
+	whd_wifi_stop_ap(airoc_ap_if);
+	data->is_ap_up = 0;
+
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+extern int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface);
+
+static int airoc_init(const struct device *dev)
+{
+	cy_rslt_t ret = CY_RSLT_SUCCESS;
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+
+	k_tid_t tid = k_thread_create(
+		&airoc_wifi_event_thread, airoc_wifi_event_stack,
+		CONFIG_AIROC_WIFI_EVENT_TASK_STACK_SIZE, (k_thread_entry_t)airoc_event_task, NULL,
+		NULL, NULL, CONFIG_AIROC_WIFI_EVENT_TASK_PRIO, K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_thread_name_set(tid, "airoc_event");
+
+	ret = airoc_wifi_init_primary(dev, &airoc_if);
+	if (ret != CY_RSLT_SUCCESS) {
+		LOG_ERR("airoc_wifi_init_primary failed ret = %d \r\n", ret);
+	}
+
+	ret = whd_management_set_event_handler(airoc_if, sta_link_events, link_events_handler, NULL,
+					       &sta_event_handler_index);
+	if (ret != CY_RSLT_SUCCESS) {
+		LOG_ERR("whd_management_set_event_handler failed ret = %d \r\n", ret);
+	}
+
+	k_sem_init(&data->sema_common, 1, 1);
+	k_sem_init(&data->sema_scan, 0, 1);
```

Comment:
k_sem_init returns a status code, so both calls above need to be checked for success.

Reply:
updated